infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/VcsSshKeysProvisioner.java

```diff
@@ -0,0 +1,218 @@
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.ssh.server.SshManager;
+import org.eclipse.che.api.ssh.server.model.impl.SshPairImpl;
+import org.eclipse.che.api.ssh.shared.model.SshPair;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.commons.annotation.Traced;
+import org.eclipse.che.commons.tracing.TracingTags;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class allows mount user SSH Keys as <a
+ * href="https://kubernetes.io/docs/concepts/configuration/secret">Kubernetes Secrets</a> on each
+ * workspace containers. <br>
+ * <strong>How it works</strong> <br>
+ * When starting a workspace, we will grep all SSH Keys registered for VCS and create K8s Secret the
+ * to them. Then secrets will be mounted on each container by path
+ * '/etc/ssh/{sshKeyName}/ssh-privatekey' as a file. <a
+ * href=https://github.com/kubernetes/kubernetes/blob/7693a1d5fe2a35b6e2e205f03ae9b3eddcdabc6b/pkg/apis/core/types.go#L4458">Secret
+ * Type for SSH keys</a> Also will be created and mounted ConfigMap with SSH settings to the
+ * '/etc/ssh/ssh_config'.
+ *
+ * @author Vitalii Parfonov
+ */
+public class VcsSshKeysProvisioner implements ConfigurationProvisioner<KubernetesEnvironment> {
+
+  // SecretTypeSSHAuth contains data needed for SSH authentication.
+  // Required field:
+  // - Secret.Data["ssh-privatekey"] - private SSH key needed for authentication
+  private static final String SECRET_TYPE_SSH = "kubernetes.io/ssh-auth";
+
+  // SSHAuthPrivateKey is the key of the required SSH private key for SecretTypeSSHAuth secrets
+  private static final String SSH_PRIVATE_KEY = "ssh-privatekey";
+
+  private static final String SSH_BASE_CONFIG_PATH = "/etc/ssh/";
+
+  private static final String SSH_CONFIG = "ssh_config";
+
+  private static final String SSH_CONFIG_PATH = SSH_BASE_CONFIG_PATH + SSH_CONFIG;
+
+  private static final Logger LOG = LoggerFactory.getLogger(VcsSshKeysProvisioner.class);
+
+  private final SshManager sshManager;
+
+  @Inject
+  public VcsSshKeysProvisioner(SshManager sshManager) {
+    this.sshManager = sshManager;
+  }
+
+  @Override
+  @Traced
+  public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+
+    TracingTags.WORKSPACE_ID.set(identity::getWorkspaceId);
+
+    StringBuilder sshConfig = new StringBuilder();
+    Map<String, String> map = new HashMap<>();
+
+    try {
+      List<SshPairImpl> sshPairs = sshManager.getPairs(identity.getOwnerId(), "vcs");
+      if (!sshPairs.isEmpty()) {
+        for (SshPair sshPair : sshPairs) {
+          if (isNullOrEmpty(sshPair.getName()) || isNullOrEmpty(sshPair.getPrivateKey())) {
+            continue;
+          }
+          Secret secret =
```

Comment:
Quite much low-level and high-level logic code are mixed here. I propose to move low-level code to private methods, like
```java
....
 public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
      throws InfrastructureException {
    TracingTags.WORKSPACE_ID.set(identity::getWorkspaceId);

    try {
      List<SshPairImpl> sshPairs = sshManager.getPairs(identity.getOwnerId(), "vcs");
      if (sshPairs.isEmpty()) {
        return;
      }

      StringBuilder sshConfig = new StringBuilder();

      for (SshPair sshPair : sshPairs) {
        doProvisionSshKey(sshPair, k8sEnv);

        sshConfig.append(buildConfig(sshPair.getName()));
      }

      doProvisionSshConfig(sshConfig.toString(), k8sEnv);
    } catch (ServerException e) {
      LOG.warn("Unable get SSH Keys. Cause: %s", e.getMessage(), e);
    }
  }

  private void doProvisionSshKey(SshPair sshPair, KubernetesEnvironment k8sEnv) {
  ...
  }

  private void doProvisionSshConfig(String sshConfig, KubernetesEnvironment k8sEnv) {
  ...
  }
```

Reply:
thanks